# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [4]:
# Your code here
import pandas as pd
import numpy as np
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
column_names = ["Customer","ST","GENDER","Education","Customer Lifetime Value",
                "Income","Monthly Premium Auto","Number of Open Complaints","Policy Type",
                "Vehicle Class","Total Claim Amount"]

In [5]:
# loading df from url can take a long time. 
# If this take a long time even after restarting kernel:
# recommend uncommenting the below command instead and commenting out the current one
#df = pd.read_csv("fallbackdata-bootcamp-v4_data_main_file1.csv", usecols=column_names)
df = pd.read_csv(url, usecols=column_names)

In [6]:
# Renaming columns
df.columns = [x.lower() for x in df.columns]
df.columns = df.columns.str.replace(' ', '_')
df.rename(columns={"st":"state"}, inplace=True)
df.columns

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [7]:
gender_dict = {'Femal':'F', 'female':'F', 'Male':'M'}
df['gender'] = df['gender'].replace(gender_dict)
#to check #df['gender'].value_counts()

In [8]:
state_dict = {'AZ':'Arizona', 'Cali':'California', 'WA':'Washington'}
df['state'] = df['state'].replace(state_dict)
#to check #df['state'].value_counts()

In [9]:
df['education'] = df['education'].replace({'Bachelors':'Bachelor'})
#to check #df['education'].value_counts()

In [10]:
car_dict = {'Sports Car':'Luxury', 'Luxury SUV':'Luxury', 'Luxury Car':'Luxury'}
df['vehicle_class'] = df['vehicle_class'].replace(car_dict)
#to check #df['vehicle_class'].value_counts()

In [11]:
df['customer_lifetime_value'] = df['customer_lifetime_value'].str.replace('%','')
df['customer_lifetime_value'].head()

0           NaN
1     697953.59
2    1288743.17
3     764586.18
4     536307.65
Name: customer_lifetime_value, dtype: object

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [12]:
df['customer_lifetime_value'] = df['customer_lifetime_value'].astype(float)
# one time operation for open complaints column
df['number_of_open_complaints'] = df['number_of_open_complaints'].str[2:-3]

In [13]:
df['number_of_open_complaints'].value_counts()

0    830
1    138
2     50
3     34
4     13
5      6
Name: number_of_open_complaints, dtype: int64

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers using `applymap()`.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     954 non-null    object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   float64
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(4), object(7)
memory usage: 344.6+ KB


### Strategy for dealing with the null values
- Gender column has the most null rows out of the dataset. The rest of the columns mostly have 1071 non null. 
- Strategy for dealing with null values is therefore using two methods
    - Fill null values in gender with the mode
    - Use dropna for the rest of the dataframe. 
        - this way the dataframe would be 100 rows larger than using dropna to begin with.

In [15]:
test_df = df
test_df['gender'].mode() # equals 'F'
test_df['gender'] = test_df['gender'].fillna('F')
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     4008 non-null   object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   float64
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(4), object(7)
memory usage: 344.6+ KB


In [16]:
test_df = test_df.dropna()
test_df.isna().sum()

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [17]:
# Number of open complaints now turned to an integer data type.
test_df['number_of_open_complaints'] = test_df['number_of_open_complaints'].astype(int)

/var/folders/y3/f9twdr796bj91_x22w2_mf2h0000gn/T/ipykernel_51492/3481935465.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_df['number_of_open_complaints'] = test_df['number_of_open_complaints'].astype(int)


## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [18]:
# After dropping duplicates from previous exercise, no duplicated rows found in the df. Will use original df.
test_df.duplicated().sum()

0

In [19]:
dup_df = df.loc[df.duplicated() == False]
dup_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1072 entries, 0 to 1071
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     1072 non-null   object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   float64
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(4), object(7)
memory usage: 100.5+ KB


In [20]:
# returning the cleaned df as a csv file
dup_df.to_csv("cleaned_df.csv")

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [4]:
import bonus_challenge_2 # local .py file. 
bonus_challenge_2.column_cleaner_pipeline()

Please enter the csv url: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


KeyboardInterrupt: 

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [21]:
import numpy as np
potential_df = test_df.loc[(test_df['total_claim_amount'] > np.quantile(test_df['total_claim_amount'], 0.75)) 
            & (test_df['customer_lifetime_value'] > np.quantile(test_df['customer_lifetime_value'], 0.75))]
potential_df.head()

customer       state gender             education  customer_lifetime_value  \
2   AI49188      Nevada      F              Bachelor               1288743.17   
31  UL64533      Nevada      F  High School or Below               1897545.61   
36  KY38074  California      F              Bachelor               1389173.57   
50  WE95729      Oregon      F               College               3670742.64   
90  JY67916      Oregon      F               College               1317101.28   

     income  monthly_premium_auto  number_of_open_complaints     policy_type  \
2   48767.0                 108.0                          0   Personal Auto   
31  65999.0                 237.0                          0  Corporate Auto   
36  42305.0                 117.0                          0   Personal Auto   
50  24804.0                 104.0                          0   Personal Auto   
90  21513.0                 119.0                          1   Personal Auto   

    vehicle_class  total_claim_amount  
2    Two-Door Car          566.472247  
31         Luxury          615.927769  
36  Four-Door Car          561.600000  
50            SUV          593.830288  
90   Two-Door Car          679.827592

In [22]:
display(test_df.describe())
display(potential_df.describe())

customer_lifetime_value        income  monthly_premium_auto  \
count             1.068000e+03   1068.000000           1068.000000   
mean              7.936903e+05  39315.411985            192.691948   
std               6.434784e+05  30484.373979           1603.241719   
min               2.004351e+05      0.000000             61.000000   
25%               4.034080e+05  14072.000000             68.000000   
50%               5.881742e+05  36226.000000             82.500000   
75%               8.962872e+05  64668.750000            109.250000   
max               5.816655e+06  99960.000000          35354.000000   

       number_of_open_complaints  total_claim_amount  
count                1068.000000         1068.000000  
mean                    0.395131          405.870113  
std                     0.884764          292.917158  
min                     0.000000            0.382107  
25%                     0.000000          204.071460  
50%                     0.000000          355.200000  
75%                     0.000000          532.800000  
max                     5.000000         2893.239678

customer_lifetime_value        income  monthly_premium_auto  \
count             1.000000e+02    100.000000            100.000000   
mean              1.482684e+06  27793.480000            142.350000   
std               7.018750e+05  26243.976583             49.605184   
min               8.997040e+05      0.000000             81.000000   
25%               1.015589e+06      0.000000            113.750000   
50%               1.271916e+06  25216.500000            124.500000   
75%               1.568487e+06  48592.250000            160.250000   
max               3.844586e+06  99002.000000            295.000000   

       number_of_open_complaints  total_claim_amount  
count                 100.000000          100.000000  
mean                    0.300000          807.113016  
std                     0.731679          305.994358  
min                     0.000000          537.600000  
25%                     0.000000          595.200000  
50%                     0.000000          670.271325  
75%                     0.000000          921.600000  
max                     4.000000         2027.724442